nucleus/common/common-util/src/main/java/com/sun/enterprise/loader/DirWatcher.java

```diff
@@ -171,15 +175,17 @@ private synchronized void syncWatches() {
             boolean valid = key.reset();
             if (!valid) {
                 processor.cancelled();
-                processors.remove(root);
             }
         }
     }
 
     private void unregisterFromFilesystem(Path path) {
-        WatchProcessor processor = processors.remove(path);
+        WatchProcessor processor = processors.get(path);
         if (processor != null) {
-            processor.cancelled();
+            if (processor.unsubscribe()) {
+                processor.cancelled();
```

Comment:
Maybe the cancel should always occur as part of `unsubscribe` when the count gets to zero?